dfc/gcp.go

```diff
@@ -194,9 +194,16 @@ func (gcpimpl *gcpimpl) listbucket(ct context.Context, bucket string, msg *GetMs
 
 		reslist.Entries = append(reslist.Entries, entry)
 	}
+
 	if glog.V(4) {
 		glog.Infof("listbucket count %d", len(reslist.Entries))
 	}
+
+	if msg.GetTargetURL {
```

Comment:
I do not think this way works. gcp & aws modules does not know where their objects are. When proxy gets a request for cloud bucket list, it asks a random target for it. So, the returned result is incorrect.
One way to fix: ask every target about their objects and add DirectURL depending on the information. It may not work well if objects are not cached yet.
Another way: get the list from one target and then at proxy side traverse the list and using hrw* function detect which target is responsible. Then add its DirectURL

Reply:
or is this case is only needed for local buckets?

Reply:
As I understand the idea, we are going to provide direct link to targets (not depending on if object is in cloud or local bucket). In all cases the responsible target is one selected by hrw*. If you ask incorrect target to get an object, in best case you get performance drop because the target will go to cloud for the object - it should not cache it because it does not belong to its cache.

Reply:
For local buckets this code and the code in aws.go are never executed

Reply:
does aws or google cloud even have a concept of which 'host/target'? is it always go to s3.aws.com?

Reply:
What about cached data? I thought the typical use-case of s3 with directurl is:  they do prefetch, so data go from cloud to local targets and cached. Then they call list with directUrls. It would be as fast as local bucket and no cloud interaction.

Otherwise I do not see sense in providing direct url for s3/aws - if they call s3/aws instead of DFC then why do they need DFC at all? :)